File: algorithm/Python3_version/leetcode673.number-of-longest-increasing-subsequence.py
class Solution:
    def findNumberOfLIS(self, nums: List[int]) -> int:
        n = len(nums)
        maxlen = 1
        dp = [1]*n
        count = [1]*n
        for i in range(1,n):
            for j in range(i):
                if nums[i] > nums[j]:
                    if dp[j]+1 > dp[i]:
                        dp[i] = dp[j]+1
                        count[i] = count[j]
                    elif dp[j]+1 == dp[i]:
                        count[i] += count[j]
            maxlen = max(maxlen, dp[i])
        ans = 0
        for i in range(n):
            if dp[i]==maxlen:
                ans += count[i]
        return ans

File: algorithm/leetcode673.number-of-longest-increasing-subsequence.cpp
class Solution {
public:
    int findNumberOfLIS(vector<int>& nums) {
    	int i, j, n = nums.size();
    	int maxlen = 1, sum = 0;
    	vector<int> dp(n, 1);
        vector<int> count(n, 1);
    	for(i = 1; i < n; ++i)
    	{
    		for(j = i-1; j >= 0; --j)
    		{
    			if(nums[i] > nums[j])
    			{
                    if(dp[j]+1 > dp[i])
                    {
                        dp[i] = dp[j]+1;
                        count[i] = count[j];
                    }
                    else if(dp[j]+1 == dp[i])
                        count[i] += count[j];
                }
    		}
            maxlen = max(maxlen, dp[i]);
    	}
        for(i = 0; i < n; ++i)
            if(dp[i]==maxlen)
                sum += count[i];
        return sum;
    }
};